### ISSUE TITLE
Add support for build-tool-depends

### ISSUE THREAD

**Issue by RyanGlScott** [sol/hpack#254]
http://cabal.readthedocs.io/en/latest/developing-packages.html?highlight=build-tool-depends#pkg-field-build-tool-depends

I found myself wanting this [here](https://github.com/haskell-compat/base-orphans/commit/7af12287c025cddab7af6f37313e98af1f3b227b), but because `hpack` doesn't support `build-tool-depends`, I have to add it to the generated `.cabal` file _post facto_.

**Comment by sol** [comment]
@RyanGlScott thank you for opening this issue.  I would like to get the full picture here first, can you give more context on how this fails.  `hspec-discover` is a transitive dependency of `hspec`.  In the past (with cabal sandboxes and `stack`) it just worked to depend on `hspec` and `hspec-discover` would be on the path when the test suite is build.

Does this still work with `cabal new-build`?

**Comment by RyanGlScott** [comment]
You have to declare a `build-tool` dependency on `hspec-discover` in order for it to build successfully with `hspec-discover` (as demonstrated [in this Travis build](https://travis-ci.org/haskell-compat/base-orphans/jobs/330394844#L770), from a commit where I forgot to add it). `new-build` uses a more declarative approach to executable dependencies than old-`build` does, and won't simply install executables as a side effect of installing the corresponding libraries anymore.

**Comment by sol** [comment]
Not excited!  Even so I understand the rational for this from the perspective of a cabal developer, this means that every user of `hspec` has to add

```
build-tool-depends: hspec-discover:hspec-discover == 2.
```

to their cabal file.

Cabal's design decision makes a lot of sense from a technical point of view, BUT it puts additional burden on the user.  A user already stated that she wants to use `hspec` and it does not make things any more convenient for her to state it multiple times.

Besides that:

- it worked in the past, so it's even harder for me to sell this as a "feature" to the average user
- it actively pushes people to requiring `cabal-version:  >= 2.0`

@snoyberg I think you have been working towards component based builds ala `cabal-2` in stack.  How will stack deal with situation like this in the future?

**Comment by sol** [comment]
@RyanGlScott An other thing I'm puzzled about, do you know if `cabal new-*` still solves the dependencies for all components as one set?  Say does

```
build-depends:
  hspec == 2.4.7
build-tool-depends:
  hspec-discover:hspec-discover
```

imply `hspec-discover == 2.4.7`?

**Comment by sol** [comment]
> install executables onto your PATH as a side effect of installing the corresponding libraries

Regarding this, do you know if the the executables are still built, but not put on the PATH, or not built at all?

**Comment by sol** [comment]
For the numbers, there are 1295 packages that depend on `hspec` on Hackage (+ an unknown list of in-house packages).

**Comment by RyanGlScott** [comment]
re https://github.com/sol/hpack/issues/254#issuecomment-358841072: No. `build-depends` and `build-tool-depends` can have independent constraints. For instance, if you have `build-depends: hspec == 2.4.7` and `build-tool-depends: hspec-discover:hspec-discover == 2.4.6`, then you'll get a build plan where `hspec-discover-2.4.7` is built as a build dependency, but `hspec-discover-2.4.6` is built as a build-tool dependency.

> Regarding this, do you know if the the executables are still built, but not put on the PATH, or not built at all?

To be precise, the executables are built in a Nix-style store, but they are only brought onto your `PATH` if they're listed as build-tool dependencies. I believe this is done to avoid clobbering any existing executables that would otherwise be present on you system, but I am not a `Cabal` developer, so this is only speculation on my part.

And FWIW, only `new-build` imposes this requirement—old-style `build` will continue to be able to use `hspec-discover` without listing it as an explicit build-tool dependency.

**Comment by sol** [comment]
> And FWIW, only new-build imposes this requirement—old-style build will continue to be able to use hspec-discover without listing it as an explicit build-tool dependency.

Yes, makes sense.  I guess my concern here is that `-new` both introduces new features (dependency caching) and at the same time redefines its view on what a valid package is.  Dependency caching on it's own is a useful feature and it would be great if you could use it with existing cabal packages (which is possible as [`sol/tinc`](https://github.com/sol/tinc) demonstrates, even without any modifications to Cabal).

Lumping features together into "one big new thing" was never an approach that I personally as able to leverage for great benefit.

**Comment by sol** [comment]
Going forward, what I'm tempted to do is add a cabal escape hatch (undocumented and not recommended to use in general).  In the past I was reluctant to add this kind of feature, as I don't want to see every other `package.yaml` using it (`hpack` should be able to handle every reasonable use case).  But it would allow you to add a temporary workaround and will help with similar situation in the future.

@RyanGlScott would that work for you?

**Comment by RyanGlScott** [comment]
Sure, sounds good to me.

**Comment by snoyberg** [comment]
Stack isn't doing anything like the new-build approach, so no changes are necessary. I also understand the motivation for this change, but like most things related to executable dependencies in Cabal: I think it's too late to make a change now.

**Comment by sol** [comment]
With what is on branch `verbatim` you can include https://github.com/sol/hpack-defaults/blob/hspec-cabal-new/.hpack/defaults.yaml like so:

```yaml
tests:
  spec:
    defaults: sol/hpack-defaults@hspec-cabal-new
    dependencies:
      - base-orphans
      - QuickCheck
```
which generates the cabal file you want.

This is a temporary workaround.  I'm still undecided on the wider issue.

**Comment by mitchellwrosen** [comment]
Hi, just as another data point I was unable to build `generics-eot` with `cabal new-build` for the same reason. It too uses a `package.yaml`.

I'm a bit confused about the rational behind not adding `build-tool-depends` to `hpack`. Let's not have some sort of contentious hard-fork from Cabal-the-file-format if we can help it ;)

**Comment by sol** [comment]
@mitchellwrosen I'm sorry for that.  `hpack-0.24.0` introduces `verbatim` as a Cabal escape hatch (check the [relevant section in the README](https://github.com/sol/hpack#passing-things-to-cabal-verbatim)).  This can be used as a workaround for now like so:

```yaml
tests:
  spec:
    main: Spec.hs
    verbatim:
      build-tool-depends:
        hspec-discover:hspec-discover == 2.*
    source-dirs: test
    dependencies:
      - hspec == 2.*
```

However, this will not guarantee that `hspec-discover` and `hspec` are of the same version + I'm not sure how you would guarantee that without requiring exact version on both `hpack` and `hspec-discover:hspec-discover`.

**Comment by sol** [comment]
Or better yet:
```yaml
tests:
  spec:
    defaults: sol/hpack-defaults@hspec-cabal-new
```

**Comment by sol** [comment]
Or if you want reproducibility (which never hurts!):
```yaml
tests:
  spec:
    defaults: sol/hpack-defaults@e2f7990163fdbcfa884a9e3202664122f9f6cd24
```
Sorry, no tag for now..

**Comment by sol** [comment]
> I'm a bit confused about the rational behind not adding build-tool-depends to hpack. Let's not have some sort of contentious hard-fork from Cabal-the-file-format if we can help it ;)

@mitchellwrosen I'm sorry if I gave the impression that we won't add support for `build-tool-depends`.  If adding `build-tool-depends` would solve the problem than I would be more than happy to do so.  But first I want to have the full picture!

Open questions:

- How to guarantee consistent versions between a library and an associated build tool
- What about compatibility with older versions of cabal? Should `build-tool-depends` imply `cabal-version: >= 2`?
- Backward compatible `.cabal` fies require a careful combination of `build-tool-depends` and `build-tools`, do we want to solve this in Hpack or leave it to the user?

**Comment by sol** [comment]
Blocked on https://github.com/haskell/cabal/issues/5105

**Comment by sol** [comment]
https://github.com/sol/hpack/pull/316 is my proposal to address this.  I'm hoping for at least one thorough review before merging that change. Any help with that is much appreciated.
